### ISSUE TITLE
Enhancement: [max-params] add option to check overload signatures as well

### ISSUE THREAD

**Issue by pillowfication** [typescript-eslint/typescript-eslint#10293]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/max-params/

### Description

Currently the `@typescript-eslint/max-params` rule only checks implementation signatures. I would like for an option to also check overload signatures which may have a different number of parameters.

### Fail

```TypeScript
// ❌ Function 'myFunc' has too many parameters (7). Maximum allowed is 4.
myFunc (a, b, c, d, e, f, g);

myFunc (...args) {
  // ...
}
```

### Pass

```TypeScript
myFunc (a, b, c);

myFunc (...args) {
  // ...
}
```

**Comment by kirkwaiblinger** [comment]
+1, seems reasonable and likely trivial to implement.

Guessing we could just add a TSDeclareFunction handler with the same impl as `FunctionDeclaration` to
https://github.com/typescript-eslint/typescript-eslint/blob/3f02687d964c1e7888e5db62256d56cd72c4f281/packages/eslint-plugin/src/rules/max-params.ts#L95-L99

**Comment by injae-kim** [comment]
Fix PR: https://github.com/typescript-eslint/typescript-eslint/pull/10312
